### ISSUE TITLE
Sound Drivers some games just do not work without them.

### ISSUE THREAD

**Issue by Akuma-Git** [dosbox-staging/dosbox-staging#613]
**Sound Drivers some games just do not work without them.**

Most people experience problems with these games have them because the game expects the driver being loaded via autoexec.bat unlike DOSBox.

Luckily someone had already done some preliminary work (see sources below) and I added some of my own findings along with it. Of course you have to deal with multiple versions of these drivers that cause all kinds of trouble. Some AdLib games need an extra parameter 'ba000'. Some games only work with an early version of a specific driver (ct-voice.drv)
IIRC one of the CT-VOICE.DRV had hardcoded ADDRESS and IRQ settings (geez), so this could be something that needs intercepting.

My workaround solution now is to have them installed to a userdrive and inside the PATH.
But some games need them in the game directory. When choosing a CT-VOICE.DRV, you need to match the SBTYPE you are using. So a SB16 version of CT-VOICE.DRV will only work 
with:
[sblaster]
sbtype  = sb16

The only exception to this rule is when the selected card is backwards compatible.
So a SB16 can work with a SB1.0 or SB2.0 version of CT-VOICE.DRV but YMMV.

```
63b9ee3aa25d2d2fcc9113cc83b7f1e8c63d95dbaebcf547537fbbdfddee3ad0  SB10/CT-VOICE.DRV
3c8497742c60cdba0cac256655e4b7cdce1c601c597395f7f5da3dfc9937079e  SB15/CT-VOICE.DRV
8af1289b6bf9d8706017fa9756cdcbee6bf3ad2d63af8d80d61798747c7bb0d1  SB20/CT-VOICE.DRV
8c988e16e220c4076484824372b7e51ff802cadd69de643ff43e6d950eb7ceae  SBPRO/1992/CT-VOICE.DRV
a8680722831059a9504077196cf76c6fa8d6f42573d3427ffeed65fcc7557229  SBPRO/1993/CT-VOICE.DRV
```

I don't know if there is a generic solution to this, 
but afaict a list of games with maybe a solution to point to would be a good start. 

**Creative Music System driver: CMSDRV.COM**
- Miami Vice - Capstone (1989)

**AdLib Driver: SOUND.COM**
- Bug Bomber (1992)
- Monty Python's Flying Circus (1990)
- Locomotion (1992)
- Punisher, The (1990)
- Rick Dangerous 2 (1991)
- Space 1889 (1990)
- Spellcasting 101: Sorcerers Get All The Girls (1990)
- The Adventures of Maddog Williams in The Dungeons of Duridian (1992)
- The Amazing Spider-Man and Captain America in Dr. Doom's Revenge! (1989)
- The Game of Harmony/E-motion (1990)
- Wild Streets (1990)
- X-Men (1989)
- X-Men II: The Fall of the Mutants (1991)

_(Creative's SB-SOUND.COM is not 100% compatible, and should be avoided)_

**Sound Blaster: CT-VOICE.DRV**:
- Elf Land - Volume One - Carl Erikson (1992)
- Eye of the Storm - Rebellion (1994)
- Elvira the Arcade Game (1991)(Flair)
- Innocent Until Caught - Divide By Zero (1993)
- Pizza Connection - Software 2000 (1995) (German original only)
- Prehistorik - Titus (1991)
- Stick Fighter 1 & 2
- Der Clou! - Neo (1994)
- Traffic Department 2192 - Safari Software (1993)

**Sources:**
https://www.vogons.org/viewtopic.php?f=33&t=76540
https://nerdlypleasures.blogspot.com/2013/12/sound-blaster-drivers-when-dos-games.html
https://www.vogons.org/viewtopic.php?f=46&t=72416&p=860775&hilit=ct%20ctvoice%20ct-voice%20voice%20drv#p860775
https://www.dosbox.com/comp_list.php

EDIT: Updated the list, added a new source.

**Comment by dreamer** [comment]
What do you mean by "unlike DOSBox"? Which game exactly shows the regression and compared to what version of DOSBox? 0.74-3 or SVN (which revision)? And which version of dosbox-staging are you testing? What conf settings are you using?

We have all changes that were introduced in SVN with the exception of ones that we know introduced regressions, and we *did not do any major reworks in the area of Adlib nor SB emulation*.

Articles you linked to refer to playing games using real hardware, not via DOSBox… As it stands right now, I don't know what this report is about, it's not actionable.

**Comment by Akuma-Git** [comment]
> What do you mean by "unlike DOSBox"? Which game exactly shows the regression and compared to what version of DOSBox? 0.74-3 or SVN (which revision)? And which version of dosbox-staging are you testing? What conf settings are you using?
> 
> We have all changes that were introduced in SVN with the exception of ones that we know introduced regressions, and we _did not do any major reworks in the area of Adlib nor SB emulation_.

It's not a regression :frowning_face: , these games always required those drivers to function.
They assume those drivers are loaded but they are not by default in DOSBox.
Most games work out of the box, but these do not.
* Prehistorik: not able to select SoundBlaster, unless a proper version of ct-voice.drv is used.
* Space: 1889: no music, if sound.com ba000 is not loaded.
* Rick Dangerous 2: unresponsive/hangs if AdLib is selected if sound.com is not loaded
* X-Men: Madness in Murderworld, falls back to speaker. is sound.com ba000 is not loaded.
etc..

(These drivers are not present on the original media, but come from the sound-card manufacturer or 3rd party.)

Configuration: defaults
Staging: v0.76.0-alpha-706-g8af36db5
SVN: 4361

> Articles you linked to refer to playing games using real hardware, not via DOSBox… As it stands right now, I don't know what this report is about, it's not actionable.

Correct, that's why I tagged this with question instead of filing bug report.
I am creating a list of games that require drivers to function properly in DOSBox.

The idea was to accumulate those from posts in this thread when people run into them.

**Comment by LowLevelMahn** [comment]
> It's not a regression  , these games always required those drivers to function.
> They assume those drivers are loaded but they are not by default in DOSBox.
> Most games work out of the box, but these do not.

so these game are using some sort of drive api? is it a limited set of available drivers out there - so maybe feasible to implement these apis - or would you load the drivers?

**Comment by kcgen** [comment]
Awesome roundup @Akuma-Git (and @rderooy, [here](https://www.vogons.org/viewtopic.php?p=894522#p894522)).

Given the drivers are single files (stand-alone) and trivial to load (ie: a single command), I think a wiki list + all-in-one zip pack that exactly parallels the GUS list and ultrasnd bundle [here](https://github.com/dosbox-staging/dosbox-staging/wiki/Gravis-Ultrasound-(GUS)) is the easiest but admittedly indirect way of helping users with this group of games. 

Going a step further so users don't even have to know about this: these games should simply be packaged the driver and the startup batch file call it if needed.  Not doing so out of some licensing concern or fear that Creative requires "separation" of the driver from the game and can only be combined by the user after hardware purchase (or some nonsense like that), only ensures that these games' Adlib audio functionality will evaporate to time and be lost to obscure knowledge.

The internals from @LowLevelMahn show these drivers offer game developers a very simple way to push audio to the sound card; so presumably emulating these wouldn't be too hard:
- Their API front ends and TSR handlers could be enumerated and we could similarly parsing the requested VOC file or accept the FM/music sequence commands
- The backend would be a bit harder, as that's what the game devs were avoiding by using these drivers: interfacing with the "hardware", which requires initialization calls followed by a more complex sequence of port/base calls to get the audio in (and at the right pace). 
- Perhaps in some cases, this emulation layer could short-circuit the backend calls to the emulated hardware all together and simply write the PCM sequence straight into dosbox's mixer channel. 

My opinion - that's a fair bit of work to avoid placing a single file in the games directory, and adding one line the startup batch file.  Hey -- If adding perfect 3dfx Voodoo and MT-32 support similarly involved placing `3dfx.com` and `roland.com` files in a games' directory.. then this would have been the go-to solution since day-1 of dosbox, and we'd all be happily playing these games ever since.

That said, if someone is driven enough to do this work so users and those packing games for archival purposes don't have to worry about this for these ~30 games - then patches are very welcome!

**Comment by Akuma-Git** [comment]
This was a hell of a job, but worth it imo. 

Sound Driver Pack: [SoundDriverPack100.zip](https://github.com/dosbox-staging/dosbox-staging/files/5252091/SoundDriverPack100.zip)

source: https://www.vogons.org/viewtopic.php?f=33&t=76607

**Comment by kcgen** [comment]
Wow; super helpful roundup Akuma!
Thumbs up for ensuring that others many experience these games' audio as the authors intended some ~30 years ago.

**Comment by LowLevelMahn** [comment]
Nice to haves:
-real git repo for the driver pack
-programming/API description for each driver(revision) (needed if dosbox-staging ever want to have these api directly emulated)

**Comment by Akuma-Git** [comment]
> -real git repo for the driver pack?

Don't know how that works with proprietary software, even if they were freely shared in the past on the creative ftp server and bbs's.

> -is is intended that CT-VOICE.DRV/ORIGINAL and IRQ7 differ not only in IRQ 05 <-> 07 changes:
> some CT-VOICE.DRV are identical but in different folders, sometimes with IRQ7 in ORIGINAL folder?

The originals came with different settings, I encountered IRQ 3,5,7.

>     1. with IRQ7 and 100% binary identical (even in timestamps)

Yes I patched the IRQ7 files while preserving the timestamps

> just one SBPRO in ORIGINAL and the IRQ7 folder would be enough
This is just an inventory round.

>     * SB1_0du
>     * SB1_5
>     * SB1_5du
>     * SB2_0
>     * Sb2_0du
>       are 100% identical in ORIGINAL and IRQ7 folder, (both folder files with IRQ7 set)?

Correct, in the IRQ7 folder all files are on IRQ7 because DOSBox's default is IRQ7

>     1. the files in ORIGINAL\SBASP16 (CT-VOICE_PATCHED.DRV and CT-VOICE.DRV) are also 100% identical
>     2. the files in ORIGINAL\SBPRO1 (CT-VOICE.DRV and "S") are also 100% identical

I forgot to delete some working files.

>     3. in CT-VOICE.DRV\INDEX.TXT (the sha256sum: 477fc2a2b9e1407a22685a2c4b10ff2a645479e8abad31c4a94377058ae70f2b is wrongly used on the IRQ 5 and the 7 version)

Correct.

> so its a mix of IRQ5 and IRQ7 in both folders...

Incorrect, its only a mix in the 'original' folder, in the IRQ7 folder all are set to IRQ7.

> -maybe merge IRQ5.ASM and IRQ7.ASM + commandline (+ 2 batch files for the IRQ)

These are there for transparency, just like the original folder.

> i personal think it would be better to have a program that copies+patches the driver into a irq7 folder on demand then preparing it beforehand - less files nearly identical files, less problems

I don't see that as being a problem. Instant use = widest adoption.
And its not the biggest package in the world ;)

> -programming/API description for each driver(revision) (needed if dosbox-staging ever want to have these api directly emulated)
> (http://www.retroarchive.org/swag/SOUND/0050.PAS.html for CT-VOICE, maybe others: http://www.retroarchive.org/swag/SOUND)
> 
> -a table how it gets loaded
> 
> for example:
> Driver 	Commandline 	Loaded as TSR 	Loaded directly by program 	embedded into program
> CT-VOICE.DRV 	? 	No 	Yes 	Yes
> CMSDRV.COM 	? 	Yes? 	No 	No
> SOUND.COM 	? 	Yes 	No 	No
> ... 	... 	... 	... 	...

These also differ per game, eg: sound.com reserves 1k of memory (iirc) unless specified otherwise. Some games need more memory and need to invoke the b0000 where the 0000 is a hexidecimal value that will default to 1k if less. The 'sound.com ba000' (40k buffer) is the one I encountered a couple of times.

A new default could be patched into 'sound.com', there is enough memory in DOSBox anyway, thoughts ?

**Comment by Akuma-Git** [comment]
Oh, I figured out the versioning in the CT-VOICE.DRV

Driver File | Address 30h 31h | Irq 32h | Version 33h 34h |Comment 
-- | -- | -- | -- | --
ORIGINAL/SB1_0/CT-VOICE.DRV|20 02|03|0d 01|1.13
ORIGINAL/SB1_0du/CT-VOICE.DRV|20 02|07|0e 01|1.14
ORIGINAL/SB1_5/CT-VOICE.DRV|20 02|07|0e 01|1.14
ORIGINAL/SB1_5du/CT-VOICE.DRV|20 02|07|0c 02|2.12
ORIGINAL/SB2_0/CT-VOICE.DRV|20 02|07|0b 02|2.11
ORIGINAL/SB2_0du/CT-VOICE.DRV|20 02|05|0c 02|2.12
ORIGINAL/SBASP16/CT-VOICE.DRV|20 02|05|05 03|3.05
ORIGINAL/SBAWE32/CT-VOICE.DRV|20 02|05|05 04|4.05
ORIGINAL/SBPRO1_1993/CT-VOICE.DRV|20 02|05|0c 02|2.12
ORIGINAL/SBPRO1/CT-VOICE.DRV|20 02|07|0c 02|2.12
ORIGINAL/SBPRO2/CT-VOICE.DRV|20 02|07|0c 02|2.12
ORIGINAL/SBPRO2du/CT-VOICE.DRV|20 02|05|0c 02|2.12

Source: https://www.vogons.org/viewtopic.php?f=33&p=895964#p895964

**Comment by kcgen** [comment]
With roughly 30 games needing these, I'd like a table with the following columns:
1. Game name (optional game version, if multiple)
1. Which addon sound file needed, exact path in driver pack (list only one file, the best or newest)
1. Where addon driver file should be placed (inside game dir or in PATH?)
1. What dosbox SB setting need changing (default is ideal, otherwise list sbtype=, irq, ...).
1. How the game should be launched (ideally no changes if game loads the driver itself, otherwise state the TSR launch command needed before the game)

@Akuma-Git, are there any other details users need to successfully use your pack with the affected games?

I will be happy to help test all these!

**Comment by LowLevelMahn** [comment]
@Akuma-Git 

>> -real git repo for the driver pack?
>Don't know how that works with proprietary software, even if they were freely shared in the past on the creative ftp server and >bbs's.

sharing is sharing, a zip file does not make is more legal :)

>>just one SBPRO in ORIGINAL and the IRQ7 folder would be enough
>This is just an inventory round.

so 5 files can be reduced down to 2 files (IRQ 5 and 7 version) with the very same result for end user

why can i choose between 
ORIGINAL\SBPRO1, ORIGINAL\SBPRO2 and IRQ7\SBPRO to get the very same driver for IRQ 7
or ORIGINAL\SBPRO1_1993, ORIGINAL\SBPRO2du to get the very same driver for IRQ 5
and why is that easier for the user? 
Is it not "history preserving" enough to state in the Readme that
there were different installation packages/game source where this exact version was found?

>>* SB1_0du
>>* SB1_5
>>* SB1_5du
>>* SB2_0
>>* Sb2_0du
>are 100% identical in ORIGINAL and IRQ7 folder, (both folder files with IRQ7 set)?
>Correct, in the IRQ7 folder all files are on IRQ7 because DOSBox's default is IRQ7

but they are also IRQ7 in ORIGINAL? so just copies

would be better to rename IRQ7 into DOSBOX - would make more sense, also if you plan to patch the sound.com default buffer size to dosbox abilities

**Comment by Akuma-Git** [comment]
@kcgen 
Ohh, its new for me too, it seems the further I dig the more I discover
:question: all over the place.

There are some good points raised in the vogons thread, re-released games (as part of a pack or budget version) could be different. Spellcasting 101: Sorcerers Get All The Girls (1990) later released on cd-rom does not require the 'sound.com' discovered by _Rob.

Also waiting on a question I asked ripsaw8080, as he stated that the version is the DSP (chip) version. But that would mean the SBPRO1/2 are identical (at least as far as the driver is concerned)

Afaict: the driver needs to match the sbtype for the best result, but if sbtype=16 the sb1.0 CT-VOICE.DRV works too, but will probably lack stereo (if the game has stereo implemented). 

Multiple factors, its driving me nuts ;)

Let's wait for the activity to drop, then I will update the pack with all new information.

**Comment by rderooy** [comment]
If I understand it correctly, CT-VOICE.DRV is just for playing VOC files. It is described as a ".VOC Memory mode driver", and I don't know if those can be stereo.

Perhaps the developer documentation can give some insight
ftp://78.46.141.148/driver/Creative/Developer/

Or wikipedia on the VOC format:
https://en.wikipedia.org/wiki/Creative_Voice_file

Obviously SOUND.COM, SB-SOUND.COM and CMSDRV.COM will all be MONO.

**Comment by Akuma-Git** [comment]
> sharing is sharing, a zip file does not make is more legal :)

Correct, but the repo can be flagged, can you look into that ?
 
> > > just one SBPRO in ORIGINAL and the IRQ7 folder would be enough
> > > This is just an inventory round.
> 
> so 5 files can be reduced down to 2 files (IRQ 5 and 7 version) with the very same result for end user

Probably just the IRQ7 if all is done. (the version discovery :bulb: shines a new light on things )

> why can i choose between
> ORIGINAL\SBPRO1, ORIGINAL\SBPRO2 and IRQ7\SBPRO to get the very same driver for IRQ 7

Because you should use the IRQ7|SBPRO version, the originals are there for people who do not trust it or who want to patch or hexedit it themselves.

> or ORIGINAL\SBPRO1_1993, ORIGINAL\SBPRO2du to get the very same driver for IRQ 5
> and why is that easier for the user?

because they should grab the one from the IRQ7 folder that is ready for use.

> Is it not "history preserving" enough to state in the Readme that
> there were different installation packages/game source where this exact version was found?
> 
> > > * SB1_0du
> > > * SB1_5
> > > * SB1_5du
> > > * SB2_0
> > > * Sb2_0du
> > >   are 100% identical in ORIGINAL and IRQ7 folder, (both folder files with IRQ7 set)?
> > >   Correct, in the IRQ7 folder all files are on IRQ7 because DOSBox's default is IRQ7
> 
> but they are also IRQ7 in ORIGINAL? so just copies

Correct.
 
> would be better to rename IRQ7 into DOSBOX - would make more sense, also if you plan to patch the sound.com default buffer size to dosbox abilities

Yes, since you are probably not the only one confused by it I will rename it .

> btw:
> Driver File 	Address 30h 31h 	Irq 32h 	Version 33h 34h 	Comment
> ORIGINAL/SBPRO1/CT-VOICE.DRV 	20 02 	07 	0c 02 	2.12
> ORIGINAL/SBPRO2/CT-VOICE.DRV 	20 02 	07 	0c 02 	2.12
> 
> isn't it a little bit misleading to show these details for 100% identical files? they will never differ in any way

They came from different sources, but they are also designated for different hardware.
So in the IRQ7 folder they are merged, in the original folder they are not.

> why is it that relevant what the source of the driver was?

Transparency

**Comment by Akuma-Git** [comment]
> If I understand it correctly, CT-VOICE.DRV is just for playing VOC files. It is described as a ".VOC Memory mode driver", and I don't know if those can be stereo.
> 
> Perhaps the developer documentation can give some insight
> ftp://78.46.141.148/driver/Creative/Developer/
> 
> Or wikipedia on the VOC format:
> https://en.wikipedia.org/wiki/Creative_Voice_file
> 
> Obviously SOUND.COM, SB-SOUND.COM and CMSDRV.COM will all be MONO.

Good point, thanks for looking into that :+1: 

I don't know either, but I could imagine that a voc is played over a left or right speaker.

**Comment by LowLevelMahn** [comment]
>>why can i choose between
>>ORIGINAL\SBPRO1, ORIGINAL\SBPRO2 and IRQ7\SBPRO to get the very same driver for IRQ 7
>
>Because you should use the IRQ7|SBPRO version, the originals are there for people who do not trust it or who want to patch or >hexedit it themselves.

so you think that naming it ORIGINAL is trustworthy enough :)

> > why is it that relevant what the source of the driver was?
> Transparency

i start to understand your point - but i still think that a port/irq set program + Readme with the original timestamps/irq/port settings would be enough for the SBPRO case

**Comment by LowLevelMahn** [comment]
> Oh, I figured out the versioning in the CT-VOICE.DRV

a german list with Soundblader DSP Versions:
http://www.amoretro.de/guides-workshops/creative-labs-sound-blaster-modellnummern
how to get the version from the hardware: https://wiki.osdev.org/Sound_Blaster_16

**Comment by dreamer** [comment]
Wait, what? Why are we talking about some auxiliary programs outside of emulator out of a sudden?

**Comment by LowLevelMahn** [comment]
> @LowLevelMahn The problem with a program to set the port/irq is that you need one for Windows, Linux and macOS and possibly more. And you know that for Windows and macOS users you need a GUI version, otherwise you will get an endless stream of help requests because nobody reads the readme.

use a DOS c compiler - like current Watcom/Digital Mars DMC/or older Borland/Turbo C,...
and use the tool from dosbox - like IRQ5/7 from Akuma (also 16bit dos executables)

it would be a command line tool that is more or less self describing

something like CNF-DRV.EXE file [irq=x] [port=y] which also shows Drv info on action

users also use UNP, LOADFIX etc. - its not super harcore - even if you are a newbie

**Comment by LowLevelMahn** [comment]
what about games that embedded the ct-voice.drv code with irq 5 set?

**Comment by Akuma-Git** [comment]
> > > why can i choose between
> > > ORIGINAL\SBPRO1, ORIGINAL\SBPRO2 and IRQ7\SBPRO to get the very same driver for IRQ 7
> > 
> > Because you should use the IRQ7|SBPRO version, the originals are there for people who do not trust it or who want to patch or >hexedit it themselves.
> 
> so you think that naming it ORIGINAL is trustworthy enough :)

That's what the source is for ;) , the original is there, so it can be compared against patched version.

**Comment by Akuma-Git** [comment]
@LowLevelMahn : please remove the abandon-ware link

**Comment by LowLevelMahn** [comment]
> > Wait, what? Why are we talking about some auxiliary programs outside of emulator out of a sudden?
> That's probably my fault for packing in the patches with the source code.
> You think I should remove them ?

dreamer is talking about the configuration program im talking about - not related to your IRQ5/IRQ7 ASM/COM files

**Comment by Akuma-Git** [comment]
> Or wikipedia on the VOC format:
> https://en.wikipedia.org/wiki/Creative_Voice_file

There is no end to this :) I just noticed that the VOC files have versions too.
Let's hope for good backwards compatibility.

**Comment by LowLevelMahn** [comment]
http://www.shikadi.net/moddingwiki/VOC_Format
how to get a bunch of VOC files that cover all these variants

**Comment by rderooy** [comment]
Have a look at:
http://www.vgmpf.com/Wiki/index.php?title=VOC

It has this interesting comment on the top of the page:
_The format was initially limited to unsigned 8-bit PCM and ADPCM data, however, as the power of sound cards increased, the format was expanded to handle 16-bit formats and a-law and u-law compressed formats._

That seems to imply that early versions of CT-VOICE.DRV may not be able to handle later versions of the format.

**Comment by LowLevelMahn** [comment]
thankfully the CT-VOICE.DRV are the smallest from Akumas DriverPack - round about 1.5-3KLoc in asm, so it thinks its really possible to analyze what VOC formats are supported by the drivers

**Comment by rderooy** [comment]
The power of Linux
```
find . -iname \*.voc -exec file "{}" \;
```
In my games directory, here are the games that use .VOC files
- Air Bucks - VOC 1.10
- Alien Legacy - VOC 1.10 & 1.20
- Beauty and the Beast - VOC 1.10
- Caesar - VOC 1.10
- Carnage - VOC 1.10
- Dragon Lore - VOC 1.10
- Duke Nukem - false positive? RIFF WAVE audio data
- Great Naval Battles 1 - VOC 1.10
- Great Naval Battles 2 - VOC 1.10
- Great Naval Battles 3 - VOC 1.10
- Legend of Kyrandia - false positive?
- Lemmings 2 - false positive?
- Mario Teaches Typing - VOC 1.10
- Millenium - VOC 1.10
- Operation Combat II: By Land, Sea & Air - VOC 1.10
- Panzer General - VOC 1.10
- Quest for Glory 4 - false positive?
- Simcity Classic - VOC 1 & VOC 1.10
- Terminator 2 The Arcade Game - VOC 1.10
- The Dark Queen of Krynn - VOC 1.10

The Linux ``file`` command will tell you if it is a "Creative Labs voice data" file and the version. I got a few hits on .VOC files that are apparently something else which I marked as "false positive?".

VOC Version 1.10 is the most common. The interesting hits are Alien Legacy and Simcity Classic as those cover version 1 (1.00?) and version 1.20.

Note this does not mean that the games use CT-VOICE.DRV they can have their own player integrated.

Ignoring the .VOC extension, I did a more extensive search through my games directory for files where the ``file`` command thinks it is a VOC file... I find a few more hits with this:

- Cool Spot (file: *.XED) - VOC 1.10
- Brix ("some" *.BRX files, others are Creative Music (CMD data files) - VOC 1.10
- Dune (file: SD5.HSQ) - VOC 1.10
- Dynablaster (*.DAT) - VOC 1.10
- Moonstone (files in the samples subdir) - VOC 1.10
- Panzer General (in addition to the VOC files, *.TLK and *.WAR) - VOC 1.10
- Sam and Max (file: testwave) - VOC 1.10
- Star Wars X-Wing (file: testwave) - VOC 1.10
- Syndicate - (file *.DAT) - VOC 1
- Ultima 7 (file *.DAT) - VOC 1.10

**Comment by LowLevelMahn** [comment]
would be also interesting to know which VOC codecs are in use

**Comment by LowLevelMahn** [comment]
> VOC Version 1.10 is the most common. The interesting hits are Alien Legacy and Simcity Classic as those cover version 1 (1.00?) and version 1.20.

seem to be a unknown version: also found in other games
https://github.com/scummvm/scummvm/blob/0051eef251409bc11463e871fcdacf75578c8a4f/audio/decoders/voc.cpp#L66

**Comment by LowLevelMahn** [comment]
Nice:
there are also CT-VOICE.DRV Headers in the (posted by rderooy) ftp://retronn.de/driver/Creative/Developer/DEVELOPR.SB/
Soundblaster Development Kit

@rderooy
i've checked the file-source code, it seems that file expected the VOC file signature always at offset 0, so you will not find more embedded VOC files

**Comment by Akuma-Git** [comment]
Tested Street Fighter II by switching sbtype's and drivers.
In conclusion: the SB1.0/SB1.5 drivers are the most compatible,
and there was a bug when testing the SB16 driver.

driver / sbtype | sb1 | sb2 | sbpro1 | sbpro2 | sb16
-- | -- | -- | -- | -- | -- 
SB10_v1.13 | Yes | Yes | Yes | Yes | Yes
SB10_v1.14 | Yes | Yes | Yes | Yes | Yes
SB15_v1.14 | Yes | Yes | Yes | Yes | Yes
SB15_v2.12 | Yes | No  | No  | No  | No
SB20_v2.11 | No  | Yes | No  | No  | No
SB20_v2.12 | No  | Yes | No  | No  | No
SBM_v2.11 | No  | Yes | No  | No  | No
SBP_SBPM_SBP2_v2.12 | No  | No  | Yes | Yes | No
SBP_v4.01 | No  | No  | No  | No  | No
SB16ASP_v3.05 | No  | No  | No  | No | Yes*1
SBAWE32_v4.05 | No  | No  | No  | No  | No

*1 sound plays too fast (high)

Source:https://www.vogons.org/viewtopic.php?f=33&t=76607&p=896311#p896311

**Comment by kcgen** [comment]
Just want to tip my hat to the work by @Akuma-Git, @LowLevelMahn , and @rderooy (alphabetic order).  

![2020-09-24_07-56](https://user-images.githubusercontent.com/1557255/94162415-866c6e00-fe3b-11ea-82f0-d6e8ead72cdd.png)

When this deep dive crew resurfaces, this is going to be the most well though-out 30-line table helping users get the most out of these games.  If we _can_ use the latest-and-greatest VOC player for some older games, this will allow the terribly rough 2, 3, 4, (and even 8) bit VOC files to be remastered into clean sounding 16-bit tracks.

(Photo credit and copyright Dan Orr)

**Comment by LowLevelMahn** [comment]
> Tested Street Fighter II by switching sbtype's and drivers.
> In conclusion: the SB1.0/SB1.5 drivers are the most compatible,
> and there was a bug when testing the SB16 driver.

using only dosbox or real hardware - this way its not clear if you expirence dosbox or real problems

**Comment by Akuma-Git** [comment]
> > Tested Street Fighter II by switching sbtype's and drivers.
> > In conclusion: the SB1.0/SB1.5 drivers are the most compatible,
> > and there was a bug when testing the SB16 driver.
> 
> using only dosbox or real hardware? this way its not clear if you expirence dosbox or real problems

Added

**Comment by LowLevelMahn** [comment]
@Akuma-Git 
found another SB 16 driver version in https://www.vogons.org/viewtopic.php?p=176069#p176069 - in the vedit.zip there is the string 'Version 3.07.03' inside

@rderooy
the vedit.zip contains an editor that seems to support all codecs

**Comment by Akuma-Git** [comment]
@LowLevelMahn 

I think you found two ;)

File|Address|Irq|Version|Status
--|--|--|--|--
CT-VOICE.16|address: 220|05|03.07|New
CT-VOICE.PRO|address: 220|07|02.10|New
CT-VOICE.SB|address: 220|07|01.14|Present

**Comment by rderooy** [comment]
I had a look at Audacity and VLC here, but neither give codec details on VOC files. The best you get is the rate being shown in Audacity.

Also Audacity does not like v1 files, but plays v1.10 and v1.20 fine. VLC seems to be fine with all three versions.

I don't really feel like writing my own parser, just to find the codec details :-)

**Comment by LowLevelMahn** [comment]
> I think you found two ;)

ftp://retronn.de/driver/Creative/Developer/DEVELOPR.SB/ contains another one after installation in SBK\DRV

containing these strings

```
CT-VOICE
CREATIVE SOUND BLASTER CARDS
Creative Loadable Voice Memory Driver 4.01. Copyright (c) Creative Technology Ltd., 1991-1993.  All rights reserved.
```

i think you will find masses of different drivers in ftp://retronn.de/driver/Creative/

**Comment by Akuma-Git** [comment]
Found the game SIEGE which also uses CT-VOICE.DRV but like Street Fighter II (SBDRIVER.DRV) it uses another name in this case: SB0.DRV SB1.DRV SB2.DRV and SB3.DRV. So besides scanning for CT-VOICE.DRV these other file names need to added when searching.

Its a process :face_with_thermometer:

**Comment by LowLevelMahn** [comment]
> Found the game SIEGE which also uses CT-VOICE.DRV but like Street Fighter II (SBDRIVER.DRV) it uses another name in this case: SB0.DRV SB1.DRV SB2.DRV and SB3.DRV. So besides scanning for CT-VOICE.DRV these other file names need to added when searching.
> 
> Its a process

and don't forget the games that embedd the drivers - the full driver as binary blob  somewhere in the game resources - i don't know any but i've read that some programs does that

**Comment by kcgen** [comment]
> and don't forget the games that embedd the drivers - the full driver as binary blob somewhere in the game resources

That's interesting from the perspective of possibly allowing folks to replace the embedded driver with a newer/better version, and then remaster any VOC audio files at a higher bit-depth. 

But given the drivers are embedded and distributed with the original game, there is no additional work or files needed for the user to get the as-designed audio working.  

So if/when we put together a list of all the affected games and what the user needs to do to grt audio working, then these would fall in the "no action needed" category.

**Comment by LowLevelMahn** [comment]
found another CT-VOICE.DRV in ftp://retronn.de/driver/Creative/SoundBlaster2.5/1/DRV001.ZIP

```
SB CT-VOICE.DRV Compatible Driver Devel
ed by Aztech Labs Inc.
Copyright (c) Aztech Lab. Inc., 1992.
```

not the same header format as original CT-VOICE.DRVs

**Comment by LowLevelMahn** [comment]
and another one in ftp://retronn.de/driver/Creative/Soundblaster16_AWE/Vibra16/Creative_Vibra_16.rar in the "Creative Vibra16 Install Disk" installation result files

Port: 220
Irq: 5
Version: 4.04

CREATIVE SOUND BLASTER CARDS
Creative Loadable Voice Memory Driver 4.04. Copyright (c) Creative Technology Ltd., 1991-1993.  All rights reserved.

**Comment by LowLevelMahn** [comment]
and another CMSDRV.com in ftp://retronn.de/driver/Creative/CMS/CMSDRV.com

seems to be 3.20**b**

**Comment by LowLevelMahn** [comment]
and new SBFMDRV.COM versions

ftp://retronn.de/driver/Creative/Creative_FTP/creative/patches/rolcmf -> Version 1.02
ftp://retronn.de/driver/Creative/Creative_FTP/creative/patches/midcmf -> Version 1.21
 ftp://retronn.de/driver/Creative/Soundblaster16_AWE/Vibra16/Creative_Vibra_16.rar in the "Creative Vibra16 Install Disk" installation result files -> Version 1.35

**Comment by kcgen** [comment]
Do we know if Akuma's excellent roundup is on the Wiki (or a link to the vogon's thread: https://www.vogons.org/viewtopic.php?f=33&t=76607&start=40)? 

Suggest we move the salient content there (or ref the vogons article) - and close this out.

**Comment by nemo93** [comment]
[Here you go](https://github.com/dosbox-staging/dosbox-staging/wiki/Sound-issues#sound-drivers). I simply can't run through the many games, files and comments. Thus I aimed for simplicity: link to the forum and to that ticket. 

Note: I'm not sure to fully understand the exact issue as I was unable to reproduce neither for Prehistorik or Moonstone (sound is fine for both). It may be that my versions are bundled with the appropriate drivers.
